File: project-05-db-migration/code-java/src/main/java/com/example/demo/service/MigrationService.java
package com.example.demo.service;

import com.example.jooq.legacy.tables.records.OrdersV1Record;
import com.example.jooq.legacy.tables.records.UsersV1Record;
import com.example.jooq.target.tables.records.MemberRecord;
import com.example.jooq.target.tables.records.PurchaseOrderRecord;
import org.jooq.DSLContext;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static com.example.jooq.legacy.Tables.ORDERS_V1;
import static com.example.jooq.legacy.Tables.USERS_V1;
import static com.example.jooq.target.Tables.MEMBER;
import static com.example.jooq.target.Tables.PURCHASE_ORDER;

@Service
public class MigrationService {

    private final DSLContext legacyDsl;
    private final DSLContext targetDsl;

    public MigrationService(@Qualifier("legacyDsl") DSLContext legacyDsl,
                            @Qualifier("targetDsl") DSLContext targetDsl) {
        this.legacyDsl = legacyDsl;
        this.targetDsl = targetDsl;
    }

    public void migrateAll() {
        migrateUsersToMembers();
        migrateOrdersToPurchaseOrders();
    }

    private void migrateUsersToMembers() {
        List<UsersV1Record> legacyUsers = legacyDsl.selectFrom(USERS_V1).fetch();

        List<MemberRecord> memberRecords = legacyUsers.stream()
                .map(user -> {
                    MemberRecord member = targetDsl.newRecord(MEMBER);
                    member.setId(user.getId()); // Keep exact ID

                    // Name splitting
                    String[] nameParts = user.getFullName().split(" ", 2);
                    member.setFirstName(nameParts[0]);
                    member.setLastName(nameParts.length > 1 ? nameParts[1] : "");

                    // Status mapping
                    String status = processStatus(user.getUserStatus());
                    member.setStatus(status);
                    
                    LocalDateTime createdAt = user.getCreatedDate();
                    member.setCreatedAt(createdAt);
                    
                    return member;
                }).collect(Collectors.toList());

        // Batch Insert with high performance
        targetDsl.batchInsert(memberRecords).execute();
    }

    private void migrateOrdersToPurchaseOrders() {
        List<OrdersV1Record> legacyOrders = legacyDsl.selectFrom(ORDERS_V1).fetch();

        List<PurchaseOrderRecord> poRecords = legacyOrders.stream()
                .map(order -> {
                    PurchaseOrderRecord po = targetDsl.newRecord(PURCHASE_ORDER);
                    po.setId(order.getId());
                    po.setMemberId(order.getUserId());
                    po.setTotalAmount(order.getOrderTotal());
                    po.setOrderState(processOrderState(order.getState()));
                    return po;
                }).collect(Collectors.toList());

        targetDsl.batchInsert(poRecords).execute();
    }

    private String processStatus(String legacyStatus) {
        if ("ACTIVE".equalsIgnoreCase(legacyStatus) || "VALID".equalsIgnoreCase(legacyStatus)) {
            return "ACTIVE";
        }
        return "INACTIVE";
    }

    private String processOrderState(String legacyState) {
        if ("PAID".equalsIgnoreCase(legacyState) || "SUCCESS".equalsIgnoreCase(legacyState)) {
            return "COMPLETED";
        }
        return "PENDING";
    }
}

File: project-05-db-migration/code-kotlin/src/main/kotlin/com/example/demo/service/MigrationService.kt
package com.example.demo.service

import com.example.jooq.legacy.tables.references.ORDERS_V1
import com.example.jooq.legacy.tables.references.USERS_V1
import com.example.jooq.target.tables.references.MEMBER
import com.example.jooq.target.tables.references.PURCHASE_ORDER
import org.jooq.DSLContext
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service

@Service
class MigrationService(
    @Qualifier("legacyDsl") private val legacyDsl: DSLContext,
    @Qualifier("targetDsl") private val targetDsl: DSLContext
) {

    fun migrateAll() {
        migrateUsersToMembers()
        migrateOrdersToPurchaseOrders()
    }

    private fun migrateUsersToMembers() {
        val legacyUsers = legacyDsl.selectFrom(USERS_V1).fetch()

        val memberRecords = legacyUsers.map { user ->
            val member = targetDsl.newRecord(MEMBER)
            member.id = user.id
            
            // Name splitting
            val nameParts = user.fullName!!.split(" ", limit = 2)
            member.firstName = nameParts[0]
            member.lastName = if (nameParts.size > 1) nameParts[1] else ""
            
            // Status mapping
            member.status = processStatus(user.userStatus)
            member.createdAt = user.createdDate
            
            member
        }

        // Batch Insert
        targetDsl.batchInsert(memberRecords).execute()
    }

    private fun migrateOrdersToPurchaseOrders() {
        val legacyOrders = legacyDsl.selectFrom(ORDERS_V1).fetch()

        val poRecords = legacyOrders.map { order ->
            val po = targetDsl.newRecord(PURCHASE_ORDER)
            po.id = order.id
            po.memberId = order.userId
            po.totalAmount = order.orderTotal
            po.orderState = processOrderState(order.state)
            po
        }

        targetDsl.batchInsert(poRecords).execute()
    }

    private fun processStatus(legacyStatus: String?): String {
        return if (legacyStatus.equals("ACTIVE", ignoreCase = true) || legacyStatus.equals("VALID", ignoreCase = true)) {
            "ACTIVE"
        } else {
            "INACTIVE"
        }
    }

    private fun processOrderState(legacyState: String?): String {
        return if (legacyState.equals("PAID", ignoreCase = true) || legacyState.equals("SUCCESS", ignoreCase = true)) {
            "COMPLETED"
        } else {
            "PENDING"
        }
    }
}
